Add `deregisterAllFonts` method.

jhuckaby opened the pull request:
Thanks for contributing!

- [X] Have you updated CHANGELOG.md?
- [X] Fixes #1762 and #1362.

This PR adds a new API called `deregisterAllFonts()`, which will deregister (unload) all fonts that were previously registered with `registerFont()`.  This can be used to free up memory, and also resolve font conflicts.  Certain font variants clash with each other when Pango has both loaded.  This API allows a long-running process to render potentially hundreds of fonts without memory leaks or font conflicts.

See previous discussion in #1762 and #1362.

Example test script with accompanying fonts available here:
https://pixlcore.com/public/67dfda0807d530f3/node-canvas-deregister-example.zip

Tested on:

- [X] macOS 10.14.6
- [X] AWS Linux 2021
- [ ] Windows 10

chearon commented on src/Canvas.cc:
```diff
@@ -768,6 +770,14 @@ NAN_METHOD(Canvas::RegisterFont) {
+	std::for_each(font_face_list.begin(), font_face_list.end(), [&](FontFace& f) {
+      deregister_font( (unsigned char *)f.file_path );
+    });
+	font_face_list.clear();
```
I think this is probably a memory leak. We need to free up the PangoFontDescriptions held by the FontFaces. It would be great if we did that in the `FontFace` destructor.

jhuckaby replied:
Yup, you're right.  So, I added a `FontFace` destructor as you requested, and freed `user_desc` and `sys_desc` if they were non-null, but this resulted in a segmentation fault.  I tried to track it down but couldn't find the cause.  I ended up just freeing the memory in the `DeregisterAllFonts` function, which works fine on both my test platforms.  I hope this is acceptable.

chearon commented on src/Canvas.cc:
```diff
@@ -768,6 +770,14 @@ NAN_METHOD(Canvas::RegisterFont) {
 }
 
+NAN_METHOD(Canvas::DeregisterAllFonts) {
+	// Unload all fonts from pango to free up memory
```
Looks like there are some tabs throughout the PR that should be replaced with spaces.

jhuckaby replied:
All fixed, thanks!

chearon reviewed (changes requested):
I'm glad you were able to make a PR! I didn't have the time to code, but will definitely support this. I see a few things that need changes.

chearon commented on src/register_font.cc:
```diff
@@ -260,3 +260,31 @@ register_font(unsigned char *filepath) {
+  if (!success) return false;
+
+  // Tell Pango to throw away the current FontMap and create a new one. This
+  // has the effect of registering the new font in Pango by re-looking up all
```
`s/registering the new font/unregistering the font`

jhuckaby replied:
Haha yup, fixed!

jhuckaby commented on src/register_font.cc:
```diff
@@ -260,3 +260,31 @@ register_font(unsigned char *filepath) {
+  success = true;
+  #endif
+
+  if (!success) return false;
```
Ah, good catch.  I added basic error handling on the calling side, so this the return value is used now.

jhuckaby commented:
Thanks for taking a look, @chearon!  I believe that I addressed all of your concerns, except for the `FontFace` destructor, which was segfaulting on me.  However, the memory leak is fixed, just in a slightly different place.

chearon commented on src/Canvas.cc:
```diff
@@ -753,6 +754,7 @@ NAN_METHOD(Canvas::RegisterFont) {
       FontFace face;
       face.user_desc = user_desc;
       face.sys_desc = sys_desc;
+      strcpy( (char *)face.file_path, (char *) *filePath );
```
`strncpy((char *)face.file_path, (char *) *filePath, 1023)`

jhuckaby replied:
Ah, good catch.  I'll make that change tonight.  Thank you!

jhuckaby replied:
Done!

chearon reviewed (commented):
Thanks. Looks good, and I'm testing it out now. One final thing, then I'll merge!

chearon commented:
Hmm, the linter failed (`npm run lint`). Apparently that check runs after approval 🥴

jhuckaby commented:
Ah crap, I'll check it out.  I forgot about the linter. 

That should do it 😊

jhuckaby commented:
Awesome!  Thank you so much!  🎉

flohall commented:
Hello, I was very happy that this change was already released, because I also have issues with changing fonts. My project uses typescript, node.js 14 and ESM modules. The linter and my IDE (IntelliJ IDEA) cannot recognize this new method. There might be something broken around the typing. I solved this for me in this way, but it can't be intended:

```
import canvas from 'canvas';
const untypedCanvas: any = canvas as any;
untypedCanvas.deregisterAllFonts();
```
